### PR TITLE
Add a toDartString extension that optionally frees native memory

### PR DESCRIPTION
**Pull request by nirinchev**

### CODE REVIEW

**Review by fealebenpae — APPROVED** [review]

**Review by nielsenko — APPROVED** [review]
This will work, but it puts the burden to remember on the caller. Down the line I would like to have an interface where we avoid first constructing a `std::string` the copying it to a `char*` only to convert it from `utf8` to `utf16`.

**Comment by nirinchev** [comment]
I think for a more robust fix, we'll need to modify the C API, which I don't think is feasible right now.

File: lib/src/native/realm_core.dart [hunk 1]
```diff
@@ -70,7 +70,7 @@ class _RealmCore {
     return _instance ??= _RealmCore._();
   }
 
-  String get libraryVersion => _realmLib.realm_get_library_version().cast<Utf8>().toDartString();
+  String get libraryVersion => _realmLib.realm_get_library_version().toDartString()!;
```
Comment (blagoev) [inline lib/src/native/realm_core.dart:73]: I suggest we keep the `cast<Utf8>`. 
This change enables toDartString on every `Pointer<Int8>` which is not a string every time. Being more explicit here is intentional.

**Review by blagoev — CHANGES_REQUESTED** [review]
It will work, but I would suggest keeping the extension on `Pointer<utf8>`. Lets keep the intentional interface.

File: lib/src/native/realm_core.dart (hunk 1 above)
Comment (nirinchev) [reply lib/src/native/realm_core.dart]: In any case, changed it to be an extension on Pointer<Utf8> - any time either approach would have saved is probably not worth debating this further.

**Review by desistefanova — APPROVED** [review]

File: lib/src/native/realm_core.dart [hunk 2]
```diff
@@ -70,7 +70,7 @@ class _RealmCore {
     return _instance ??= _RealmCore._();
   }
 
-  String get libraryVersion => _realmLib.realm_get_library_version().cast<Utf8>().toDartString();
+  String get libraryVersion => _realmLib.realm_get_library_version().cast<Utf8>().toRealmDartString()!;
```
Comment (blagoev) [inline lib/src/native/realm_core.dart:73]: why not use `toDartString` here

Comment (blagoev) [reply lib/src/native/realm_core.dart:73]: and on all places where we use `!` in `toRealmDartString()!`

File: lib/src/native/realm_core.dart [hunk 3]
```diff
@@ -81,7 +81,7 @@ class _RealmCore {
 
     String? message;
     if (error.ref.message != nullptr) {
-      message = error.ref.message.cast<Utf8>().toDartString();
+      message = error.ref.message.cast<Utf8>().toRealmDartString();
```
Comment (blagoev) [inline lib/src/native/realm_core.dart:84]: we are not reaping the benefit of introducing `toRealmDartString` here. I think the previous version was removing the null check

File: lib/src/native/realm_core.dart [hunk 4]
```diff
@@ -81,7 +81,7 @@ class _RealmCore {
 
     String? message;
     if (error.ref.message != nullptr) {
```
Comment (blagoev) [inline lib/src/native/realm_core.dart:83]:
```suggestion
    
```

**Review by blagoev — CHANGES_REQUESTED** [review]
some changes suggested

File: lib/src/native/realm_core.dart (hunk 2 above)
Comment (nirinchev) [reply lib/src/native/realm_core.dart:73]: My thinking was that it's better to have a single method for accessing strings. That way it's less cognitive load and we'll have an easier time refactoring stuff if necessary. Getting used to using `toRealmDartString` will also remind people to check whether they need to free the native string.

File: lib/src/native/realm_core.dart (hunk 3 above)
Comment (nirinchev) [reply lib/src/native/realm_core.dart:84]: yeah, I overlooked that

**Review by blagoev — APPROVED** [review]

**Pull request merged**